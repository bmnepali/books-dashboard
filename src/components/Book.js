import React from 'react';
import PropTypes from 'prop-types';

const Book = ({ book }) => {
  const bookCoverPath = 'https://s1.nyt.com/du/books/images/';
  const bookDetails = book.book_details[0];
  const bookCoverImage = bookDetails.primary_isbn13 ? bookCoverPath + bookDetails.primary_isbn13 + '.jpg' : '';
  const fallBackImage = 'https://s3-us-west-2.amazonaws.com/s.cdpn.io/387928/book%20placeholder.png';

  return (
    <article className="media book-card">
      <object data={bookCoverImage} type="image/jpeg">
        <img src={fallBackImage} alt={bookDetails.title} />
      </object>
      <div className="media-body">
        <h5 className="mt-1">{bookDetails.title}</h5>
        <p>{bookDetails.description}</p>
        <div className="action-bar">
          <a href="#" className="btn btn-rounded btn-secondary">
            Read More
          </a>
          <span className="share-book">
            <i className="icon ion-md-share" />
          </span>
        </div>
      </div>
    </article>
  );
};

Book.propTypes = {
  book: PropTypes.shape({
    book_details: PropTypes.arrayOf(PropTypes.object),
  }).isRequired,
};

export default Book;
