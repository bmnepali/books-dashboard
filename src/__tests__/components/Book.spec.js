import * as React from 'react';

// Load Component
import Book from '../../components/Book';

describe('Book Component', () => {
  describe('when no props are provided', () => {
    it('renders Book component with an error', () => {
      expect(() => shallow(<Book />)).toThrow();
    });
  });

  describe('when expected props are provided', () => {
    const props = {
      book: {
        book_details: [
          {
            title: 'THE NEW GIRL',
            description: 'Gabriel Allon, the chief of Israeli intelligence, partners with the crown prince of Saudi Arabia, whose daughter is kidnapped.',
            contributor: 'by Daniel Silva',
            author: 'Daniel Silva',
            contributor_note: '',
            price: 0,
            age_group: '',
            publisher: 'Harper',
            primary_isbn13: '9780062834836',
            primary_isbn10: '0062834835'
          }
        ]
      }
    };

    it('renders Book component without anyout an any error', () => {
      shallow(<Book {...props}/>);
    });

    it('matches snapshot', () => {
      const wrapper = mount(<Book {...props}/>);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    describe('checking Boook component closely', () => {
      let component = null;

      beforeEach(() => {
        component = mount(<Book {...props} />);
      });

      it('renders the book cover correctly', () => {
        expect(component.find('object').length).toBe(1);
        expect(component.find('img').length).toBe(1);
        expect(component.find('object').getElement().props.data).toEqual('https://s1.nyt.com/du/books/images/9780062834836.jpg');
        expect(component.find('img').getElement().props.src).toEqual('https://s3-us-west-2.amazonaws.com/s.cdpn.io/387928/book%20placeholder.png');
      });

      it('renders the book title correctly', () => {
        expect(component.find('h5.mt-1').length).toBe(1);
        expect(component.find('h5.mt-1').text()).toEqual(props.book.book_details[0].title);
      });

      it('renders the book description correctly', () => {
        expect(component.find('p').length).toBe(1);
        expect(component.find('p').text()).toEqual(props.book.book_details[0].description);
      });
    });
  });
});
