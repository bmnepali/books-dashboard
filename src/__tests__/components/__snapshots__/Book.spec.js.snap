// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Component when expected props are provided matches snapshot 1`] = `
<Book
  book={
    Object {
      "book_details": Array [
        Object {
          "age_group": "",
          "author": "Daniel Silva",
          "contributor": "by Daniel Silva",
          "contributor_note": "",
          "description": "Gabriel Allon, the chief of Israeli intelligence, partners with the crown prince of Saudi Arabia, whose daughter is kidnapped.",
          "price": 0,
          "primary_isbn10": "0062834835",
          "primary_isbn13": "9780062834836",
          "publisher": "Harper",
          "title": "THE NEW GIRL",
        },
      ],
    }
  }
>
  <article
    className="media book-card"
  >
    <object
      data="https://s1.nyt.com/du/books/images/9780062834836.jpg"
      type="image/jpeg"
    >
      <img
        alt="THE NEW GIRL"
        src="https://s3-us-west-2.amazonaws.com/s.cdpn.io/387928/book%20placeholder.png"
      />
    </object>
    <div
      className="media-body"
    >
      <h5
        className="mt-1"
      >
        THE NEW GIRL
      </h5>
      <p>
        Gabriel Allon, the chief of Israeli intelligence, partners with the crown prince of Saudi Arabia, whose daughter is kidnapped.
      </p>
      <div
        className="action-bar"
      >
        <a
          className="btn btn-rounded btn-secondary"
          href="#"
        >
          Read More
        </a>
        <span
          className="share-book"
        >
          <i
            className="icon ion-md-share"
          />
        </span>
      </div>
    </div>
  </article>
</Book>
`;
